lib/active_merchant/billing/gateways/checkout_v2.rb

```diff
@@ -177,23 +201,31 @@ def add_payment_method(post, payment_method, options)
         end
       end
 
-      def add_customer_data(post, options)
+      def add_customer_data(post, options, billing_address_key=nil)
         post[:customer] = {}
         post[:customer][:email] = options[:email] || nil
         post[:customer][:name] = options[:name]
         post[:payment_ip] = options[:ip] if options[:ip]
+
+        add_address(post, options, billing_address_key)
+      end
+
+      def add_address(post, options, billing_address_key=nil)
         address = options[:billing_address]
         if address
-          post[:customer][:name] = options[:billing_address][:name]
-          billing_address = {}
-          billing_address[:billing_address] = {}
+          post[:customer][:name] = options[:billing_address][:name] if post[:customer]
+          billing_address = { billing_address: {} }
           billing_address[:billing_address][:address_line1] = address[:address1] unless address[:address1].blank?
           billing_address[:billing_address][:address_line2] = address[:address2] unless address[:address2].blank?
           billing_address[:billing_address][:city] = address[:city] unless address[:city].blank?
           billing_address[:billing_address][:state] = address[:state] unless address[:state].blank?
           billing_address[:billing_address][:country] = address[:country] unless address[:country].blank?
           billing_address[:billing_address][:zip] = address[:zip] unless address[:zip].blank?
-          post[:source] ? post[:source].merge!(billing_address) : post.merge!(billing_address)
+          if billing_address_key
+            (post[billing_address_key] ||= {}).merge!(billing_address)
```

Comment:
I think that this version is much more readable
```ruby
post[billing_address_key] ||= {}
post[billing_address_key].merge!(billing_address)
```
What do you think?

Reply:
I just changed it.